sources/core/Stride.Core/Diagnostics/Profiler.cs

```diff
@@ -56,37 +58,119 @@ namespace Stride.Core.Diagnostics
+            }
+
+            private Channel<ProfilingEvent> _channel;
+
+            internal ChannelWriter<ProfilingEvent> Writer => _channel.Writer;
+            internal ChannelReader<ProfilingEvent> Reader => _channel.Reader;
+        }
+
+        private class ThreadEventCollection
+        {
+            //TODO (perf): Use an SPSC queue instead.
+            private ProfilingEventChannel channel = ProfilingEventChannel.Create(new UnboundedChannelOptions { SingleReader = true, SingleWriter = true });
 
-            public RollingBuffer()
+            internal ThreadEventCollection()
             {
-                for (int i = 0; i < RollingBufferCount; i++)
-                    rollingBuffers.Add(new FastList<ProfilingEvent>());
+                Profiler.AddThread(this);
             }
 
-            public FastList<ProfilingEvent> GetBuffer()
+            internal void Add(ProfilingEvent e)
             {
-                return rollingBuffers[currentBufferIndex];
+                if (subscriberChannels.Count > 0)
+                    channel.Writer.TryWrite(e);
             }
 
-            public void MoveToNextBuffer()
+            internal IAsyncEnumerable<ProfilingEvent> ReadEvents()
             {
-                currentBufferIndex = (currentBufferIndex + 1) % RollingBufferCount;
-                rollingBuffers[currentBufferIndex].Clear();
+                return channel.Reader.ReadAllAsync();
             }
         }
 
+        //TODO: Hack. No guaranteees this won't collide with a CPU thread id.
+        //      Fix when taking another look at GPU profiling.
+        internal const int GpuThreadId = -1;
+
         internal static Logger Logger = GlobalLogger.GetLogger("Profiler"); // Global logger for all profiling
-        
-        private static readonly RollingBuffer CpuEvents = new RollingBuffer();
-        private static readonly RollingBuffer GpuEvents = new RollingBuffer();
+        internal static TimeSpan StartTime = TimeSpanFromTimeStamp(Stopwatch.GetTimestamp());
+        internal static TimeSpan GpuStartTime = TimeSpan.Zero;
+
         private static readonly object Locker = new object();
         private static bool enableAll;
         private static int profileId;
+        private static ThreadLocal<ThreadEventCollection> events = new(() => new ThreadEventCollection(), true);
+        private static ProfilingEventChannel collectorChannel = ProfilingEventChannel.Create(new UnboundedChannelOptions { SingleReader = true });
+        private static SemaphoreSlim subscriberChannelLock = new SemaphoreSlim(1, 1);
+        private static List<Channel<ProfilingEvent>> subscriberChannels = new();
+        private static Task collectorTask = null;
+
+        //TODO: Use TicksPerMicrosecond once .NET7 is available
+        public static TimeSpan MinimumProfileDuration { get; set; } = new TimeSpan(TimeSpan.TicksPerMillisecond / 1000);
+
+        public static void Init()
```

Comment:
I'm wondering if having those `Init` methods make sense.
What is the cost of running initialization in a static constructor? A new task will be created and then put to sleep until an event is produced, right?

Also the `Init(TimeSpan)` overload doesn't make much sense to me if the user can anyways set the static `MinimumProfileDuration` at any point and not only at initialization.

Reply:
Agreed, these just carried over from the initial prototype. And I think at least for now, reinitialization should not be needed.